engine/src/xetex_opentype_math.rs

```diff
@@ -1332,26 +1332,24 @@ shall not be used in advertising or otherwise to promote the sale,
 use or other dealings in this Software without prior written
 authorization from the copyright holders.
 \****************************************************************************/
-#[no_mangle]
-pub unsafe extern "C" fn get_ot_math_constant(
-    mut f: libc::c_int,
-    mut n: libc::c_int,
-) -> libc::c_int {
+pub fn get_ot_math_constant(f: i32, n: i32) -> i32 {
```

Comment:
I feel this function is lying by claiming to be safe

Reply:
`XeTeXFontInst_getHbFont` is not safe so have to be in a unsafe block. I will try to replace all the unsafe api to real safe api later.

Reply:
(Forgive me if you know that already, I don’t know how much you know. I hope it would at least guide other potential contributors.)

Reply:
Yes, access a static mut `font_area` is not safe, so has to be in a unsafe block. but changing from 

```rust
pub static mut font_area: *mut str_number = 0 as *const str_number as *mut str_number;
```

to 

```rust
pub static mut font_area: [i32; 9001]
```
is much rusty even though essentially they are the same(both unsafe).

`font_size` is similiar to `font_area`, I don't know if I should continue turning these pointer to array, it's a lot of boring work.

Reply:
Not an unsafe block. An unsafe function modifier. Until you’ve analysed every use. An unsafe block is for when you know exactly how it behaves and can prove it’s safe. I haven’t seen that.

I would like to see every font-related static access encapsulated in a struct with unsafe functions on it. The struct should probably be behind a mutex at least to begin with. Then you can start absorbing functionality into the struct, and at some point make the functions safe by changing the statics to something else. That way you separate the rote refactoring from the safety logic.

Reply:
> it’s a lot of boring work

Welcome to the party, pal. A good IDE or `cargo install fastmod` can work wonders.

Reply:
Encapsulate font related code into a struct is be an ambitious goal, the original `xetex` is writted in C which is procedure code is hard to refactor into OOP code. I suggest we clean up the generated C like rust code first.

Reply:
It’s not *that* OOP. Doesn’t need to be a struct either, you can create a couple of functions that do the static accesses. If you’re going to touch all those lines of code, you may as well give yourself a springboard to make it all safe at once rather than have to touch them again.

Reply:
I'm thinking an evolution like this, for font_area:

```rust
unsafe fn font_format(font_id: i32) -> u32 {
    *font_area.offset(font_id as isize) as u32
}
```

Then

```rust
enum FontFormat {
    #[cfg(target_os = "macos")]
    AAT,
    OpenType,
}
#[inline]
unsafe fn font_format(font_id: i32) -> FontFormat {
    match font_area[font_id as usize] {
        #[cfg(target_os = "macos")]
        0xffffu32 => Some(FontFormat::AAT),
        0xfffeu32 => Some(FontFormat::OpenType),
        _ => panic!("invalid font format in font_area table"),
    }
}
```

Eventually

```rust
struct FontId(u32);
enum CachedFont {
    AAT { ... },
    OpenType { ... },
}
impl CachedFont {
    fn format() -> FontFormat {
        match self { ... }
    }
}
lazy_static! {
    static ref SHARED_FONT_CACHE: RwLock<SharedFontCache> = ...;
};
struct SharedFontCache {
    fonts: Vec<CachedFont>,
}
impl SharedFontCache {
    fn read_only() -> LockResult<RwLockReadGuard<SharedFontCache>> {
        SHARED_FONT_CACHE.read()
    }
    fn get_font(&self) -> &CachedFont { ... }
}
// And replace this with plain get_font as you almost always go on to access the rest of the font information
fn font_format(font_id: FontId) -> FontFormat {
    let cache = SharedFontCache::read_only();
    let font = cache.get_font(font_id);
    font.format()
}
```